hash-parser/src/emit.rs

```diff
@@ -95,23 +101,287 @@ impl ast::IntoAstNode<ast::Type> for HashPair<'_> {
+
+                        // firstly, take care of the function parameters...
+                        let args = components
+                            .next()
+                            .unwrap()
+                            .into_inner()
+                            .map(|param| {
+                                let mut param_components = param.into_inner();
+
+                                // get the name of identifier
+                                let name = param_components.next().unwrap().into_ast();
+
+                                // if no type is specified for the param, we just set it to none
+                                let ty = param_components.next().map(|t| t.into_ast());
+
+                                ab.node(FunctionDefArg { name, ty })
+                            })
+                            .collect();
+
+                        // now check here if the next rule is either a type or a expression,
+                        // if it is a type, we expect that there are two more rules to follow
+                        // since function literals cannot be without a fn_body
+                        let fn_type_or_body = components.next().unwrap();
+
+                        let (return_ty, fn_body) = match fn_type_or_body.as_rule() {
+                            Rule::any_type => {
+                                let body = components.next().unwrap();
+                                (Some(fn_type_or_body.into_ast()), body.into_ast())
+                            }
+                            Rule::expr => (None, fn_type_or_body.into_ast()),
+                            _ => unreachable!(),
+                        };
+
+                        ab.node(Literal::Function(FunctionDef {
+                            args,
+                            return_ty,
+                            fn_body,
+                        }))
+                    }
+                    _ => unreachable!(),
+                }
+            }
+            _ => unreachable!(),
+        }
+    }
+}
+
+impl IntoAstNode<LiteralPattern> for HashPair<'_> {
+    fn into_ast(self) -> AstNode<LiteralPattern> {
+        // match self.as_rule() {
+        //
+        // }
+        unimplemented!()
+    }
+}
+
+impl IntoAstNode<Pattern> for HashPair<'_> {
+    fn into_ast(self) -> AstNode<Pattern> {
+        // match self.as_rule() {
+        //
+        // }
+        unimplemented!()
+    }
+}
+
+impl IntoAstNode<Expression> for HashPair<'_> {
+    fn into_ast(self) -> AstNode<Expression> {
+        match self.as_rule() {
+            Rule::expr => {
+                let ab = AstBuilder::from_pair(&self);
+                let expr = self.into_inner().next().unwrap();
+
+                match expr.as_rule() {
+                    Rule::block => ab.node(Expression::Block(expr.into_ast())),
+                    Rule::struct_literal => ab.node(Expression::LiteralExpr(expr.into_ast())),
```

Comment:
Could do `ab.node(Expression::LiteralExpr(expr.into_ast().body))` to avoid extra AstNode, but I think this is fine as well.